cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -485,6 +644,37 @@ ArrayKernelExec ArithmeticExecFromOp(detail::GetTypeId get_id) {
   }
 }
 
+// For kernels that always return floating results
+template <template <typename... Args> class KernelGenerator, typename Op>
+ArrayKernelExec IntToDoubleExecFromOp(detail::GetTypeId get_id) {
```

Comment:
The variety of generator dispatchers available in the compute kernel are not consistent with their names. At some point we should rename them consistently, but that is out-of-scope for this PR. Nevertheless, I suggest to change the name of this generator dispatcher to `GenerateUnaryArithmeticFloatOutType`. This is simply a suggestion, we can rename later.